hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/FileGroupReaderBasedMergeHandle.java

```diff
@@ -90,7 +147,15 @@ public FileGroupReaderBasedMergeHandle(HoodieWriteConfig config, String instantT
     this.maxInstantTime = maxInstantTime;
     this.keyToNewRecords = Collections.emptyMap();
     this.readerContext = readerContext;
-    this.operation = operation;
+    this.operation = Option.of(operation);
+    initRecordTypeAndCdcLogger(enginRecordType);
+    init(operation, this.partitionPath);
+    this.props = TypedProperties.copy(config.getProps());
+    this.isCompaction = true;
```

Comment:
we already have flag `preserveMetadata` to distinguish table service and regular write, can we continue to use that? some functions like SI tracing already relies on the flag `preserveMetadata`.